lib/src/dag_set_scheduler.rs

```diff
@@ -196,20 +201,49 @@ pub trait DAGSetSchedulerBase<T: ProcessorBase + Clone> {
                 });
             }
 
-            // Allocate the nodes of ready_queue to idle cores
+            // Allocate nodes as long as there are idle cores, and attempt to preempt when all cores are busy.
```

Comment:
The following implementations are clearer:

```
            // Allocate nodes as long as there are idle cores, and attempt to preempt when all cores are busy.
            while !ready_queue.is_empty() {
                let processor = self.get_processor_mut();

                if let Some(idle_core_index) = processor.get_idle_core_index() {
                    // Allocate the node to the idle core
                    let node_data = ready_queue.pop_first().unwrap().convert_node_data();
                    self.allocate_node(
                        &node_data,
                        idle_core_index,
                        managers[node_data.get_params_value("dag_id") as usize].get_release_count()
                            as usize,
                    );
                } else if let PreemptiveType::Preemptive {
                    key: preemptive_key,
                } = &preemptive_key
                {
                    // If all cores are busy and a preemptive_key exists, attempt preemption
                    let (max_value, core_index) =
                        processor.get_max_value_and_index(preemptive_key).unwrap();
                    let ready_node_data_value = ready_queue
                        .first()
                        .unwrap()
                        .convert_node_data()
                        .get_params_value(preemptive_key);

                    if max_value > ready_node_data_value {
                        // Preempt the node with the lowest priority
                        let suspended_node_data = processor.suspend_execution(core_index).unwrap();
                        processor.allocate_specific_core(
                            core_index,
                            &ready_queue.pop_first().unwrap().convert_node_data(),
                        );
                        ready_queue.insert(NodeDataWrapper {
                            node_data: suspended_node_data,
                        });
                    }
                } else {
                    break; // No core is idle and no preemption possible, exit the loop.
                }
            }
```

Reply:
Thanks for the refactoring.

I added a break to exit the loop when no preemptive happens with all cores busy and the preemptive type specified.